### PR TITLE
Feat(MicroSplit API): HDN and MicroSplit lightning modules

### PR DESCRIPTION
**Pull request by veegalinova**
## Disclaimer

- [ ] I am an AI agent.
- [x] I have used AI and I thoroughly reviewed every line.
- [ ] I have not used AI extensively.

## Description

> [!NOTE]  
> **tldr**: This is a draft PR introducing separate HDN and MicroSplit lightning modules in the current "next-gen" format.  

### Overview - what changed?

- Added HDN and MicroSplit lightning modules, based on the old implementations and cleaned up
- Added a configuration factory (for now only HDN) to test the training performance (`create_hdn_config` / `create_advanced_hdn_config`). I tried to set "reasonable" default parameters from the paper and from our discussions with the research team.
- Light parameter cleanup based on the discussions - removed `kl_annealing`, `analytical_kl`, `KLLossConfig`, some unused code in the model and some tests

### Implementation - how did you implement the changes?

HDN now automatically determines if noise model is passed or not. If it is passed - use as is, if not - learn using `predict_logvar`. 
MMSE prediction not implemented - as Milly suggested, should be outside of the lightning module's responsibility. 

### New features or files

- `HDNModule` in `lightning/modules/hdn_module.py`
- `MicroSplitModule` in `lightning/modules/microsplit_module.py`
- `create_hdn_config` / `create_advanced_hdn_config` in `config/factories/hdn_factory.py`
### Modified features or files

- `LVAEConfig`: exposed `encoder/decoder_blocks_per_layer` parameter
- `losses/lvae`: dropped KL annealing (`kl_params`)

## How has this been tested?

Currently doing training runs on BSD. So far HDN setup runs without problems, but metrics are considerably lower than paper: 26.90 +/- 2.26 vs reported 27.42 for DivNoising and 28.82 for HDN. 
Will run a few more attempts with different parameters and update here. 

---

**Please ensure your PR meets the following requirements:**

- [ ] Code builds and passes tests locally, including doctests
- [ ] New tests have been added (for bug fixes/features)
- [ ] Documentation has been updated
- [x] Pre-commit passes

### CODE REVIEW

**Comment by jdeschamps** [comment]
Closes https://github.com/CAREamics/careamics/issues/973

File: src/careamics/compat/config/configuration_factories.py [hunk 1]
```diff
@@ -1436,8 +1435,6 @@ def _create_vae_configuration(
         Type of nonlinearity function to use.
     predict_logvar : Literal # TODO needs review
         _description_.
-    analytical_kl : bool # TODO needs clarification
-        _description_.
```
Comment (jdeschamps) [inline src/careamics/compat/config/configuration_factories.py]: Can we keep some notes somewhere in the code base to justify removing these parameters compared to previous implementations? For instance as comments in the MSplit algorithm configuration or main MSplit config.

I am afraid the knowledge will be lost.

File: src/careamics/compat/config/configuration_factories.py
Comment (jdeschamps) [inline src/careamics/compat/config/configuration_factories.py:1]: This file should be removed from this PR. The modification and the comment is still relevant for https://github.com/CAREamics/careamics/pull/1007.

Maybe this change should happen in another PR.

File: src/careamics/config/algorithms/hdn_algorithm_config.py [hunk 2]
```diff
@@ -37,6 +38,8 @@ class HDNAlgorithm(VAEBasedAlgorithm):
 
     is_supervised: bool = False
 
+    optimizer: OptimizerConfig = OptimizerConfig(name="Adamax")
```
Comment (jdeschamps) [inline src/careamics/config/algorithms/hdn_algorithm_config.py:42]:
```suggestion
    # overwrite VAE config default optimizer
    optimizer: OptimizerConfig = OptimizerConfig(name="Adamax")
```

File: src/careamics/config/architectures/lvae_config.py [hunk 3]
```diff
@@ -41,7 +45,31 @@ class LVAEConfig(ArchitectureConfig):
 
     predict_logvar: bool = True
     """Whether to predict log-variance (pixelwise uncertainty)."""
-    analytical_kl: bool = Field(default=False)
+
+    @model_validator(mode="after")
+    def validate_n_filters(self: Self) -> Self:
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_config.py]: If they are enforced to be the same, then why have two parameters and not just a single one?

File: src/careamics/config/factories/hdn_factory.py
Comment (jdeschamps) [inline src/careamics/config/factories/hdn_factory.py:1]: Merge conflicts ahead with https://github.com/CAREamics/careamics/pull/1007, it may be worth have another PR for that.

File: src/careamics/config/losses/loss_config.py [hunk 4]
```diff
@@ -5,21 +5,6 @@
 from pydantic import BaseModel, ConfigDict
 
 
-class KLLossConfig(BaseModel):
```
Comment (jdeschamps) [inline src/careamics/config/losses/loss_config.py:8]: Could you comment on why the `current_epoch` is not relevant / unused?

File: src/careamics/lightning/modules/hdn_module.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+    """CAREamics PyTorch Lightning module for the HDN algorithm.
+
+    HDN (Hierarchical DivNoising) is a VAE-based denoising algorithm built on the
+    LVAE model. It is self-supervised by default (the input patch is used as its own
+    target), but a supervised target can be provided through the second batch element.
+
+    The reconstruction likelihood is selected automatically from the configuration:
+
+    - if a noise model is provided, the noise model likelihood is used
+      (``predict_logvar`` must be ``False``);
```
Comment (jdeschamps) [inline src/careamics/lightning/modules/hdn_module.py]: Minor, but the rest of the codebase uses single apostrophe

File: src/careamics/lightning/modules/hdn_module.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        )
+
+    def on_fit_start(self) -> None:
+        """On fit start hook for HDN module."""
+        if self.noise_model is None:
+            return
+        assert self._trainer is not None
+        datamodule: CareamicsDataModule = self._trainer.datamodule  # type: ignore[union-attr]
+        normalization = cast(
+            MeanStdNormalization,
```
Comment (jdeschamps) [inline src/careamics/lightning/modules/hdn_module.py]: Why does it have to be MeantStd?

File: src/careamics/lightning/modules/hdn_module.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        batch_idx : int
+            The index of the current batch in the training loop.
+
+        Returns
+        -------
+        torch.Tensor or None
+            The loss value for the current training step, or ``None`` to skip the
+            batch when the loss is NaN.
+        """
+        x_data = cast(torch.Tensor, batch[0].data)
```
Comment (jdeschamps) [inline src/careamics/lightning/modules/hdn_module.py]: why is there a need to `cast` the data?

File: src/careamics/lightning/modules/hdn_module.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        )
+
+        return ImageRegionData(
+            data=denormalized_output,
+            source=x.source,
+            data_shape=x.data_shape,
+            dtype=x.dtype,
+            axes=x.axes,
+            region_spec=x.region_spec,
+            additional_metadata={},
```
Comment (jdeschamps) [inline src/careamics/lightning/modules/hdn_module.py]: Additional metadata must be forwarded (zarr sources).

```suggestion
            additional_metadata=x.additional_metadata,
```

File: src/careamics/lightning/modules/microsplit_module.py [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+            A tuple containing the input data and the target data.
+        batch_idx : int
+            The index of the current batch in the training loop.
+
+        Returns
+        -------
+        torch.Tensor or None
+            The loss value, or ``None`` to skip the batch when the loss is NaN.
+        """
+        x_data = cast(torch.Tensor, batch[0].data)
```
Comment (jdeschamps) [inline src/careamics/lightning/modules/microsplit_module.py]: Same question here and for all the other casts

File: src/careamics/lightning/modules/vae_lightning_module.py
Comment (jdeschamps) [inline src/careamics/lightning/modules/vae_lightning_module.py:1]: This file should be removed (it is removed in `dev/microspliut_api` and depends on `careamics.compat`)

**Review by jdeschamps — CHANGES_REQUESTED** [review]
Thanks! I didn't see much to change, just curious about a few things.

File: src/careamics/compat/config/configuration_factories.py (hunk 1 above)
Comment (CatEek) [reply src/careamics/compat/config/configuration_factories.py]: https://app.notion.com/p/HDN-usability-discussion-2b7a07b1e30c809c971bc7cd490445fe

Some of it was mentioned there, removing was Ashesh's decision

File: src/careamics/lightning/modules/microsplit_module.py [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+        Parameters
+        ----------
+        algorithm_config : VAEBasedAlgorithm or dict
+            Configuration for the MicroSplit algorithm, either as a
+            ``VAEBasedAlgorithm`` instance or a dictionary.
+        """
+        super().__init__()
+
+        if isinstance(algorithm_config, dict):
+            config = VAEBasedAlgorithm(**algorithm_config)
```
Comment (melisande-c) [inline src/careamics/lightning/modules/microsplit_module.py:62]: Just a note that in future we will separate the `VAEBasedAlgorithm` into `MicroSplitAlgorithm` and `HDNAlgorithm` (shouldn't be addressed here)

File: src/careamics/lightning/modules/microsplit_module.py [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+        x_data = cast(torch.Tensor, x.data)
+
+        # reconfigure the model for the current input spatial size
+        n_spatial_dims = x_data.dim() - 2
+        self.model.reset_for_inference(tuple(x_data.shape[-n_spatial_dims:]))
+
+        prediction = self._get_reconstruction(self.model(x_data))
+
+        denorm = Denormalize(image_means=self.target_means, image_stds=self.target_stds)
+        denormalized_output = denorm(patch=prediction.detach().cpu().numpy())
```
Comment (melisande-c) [inline src/careamics/lightning/modules/microsplit_module.py]: I guess this was present in the old version of the module, but I guess we should denormalise with the new normalisation functions.

In the CARE module we do:
https://github.com/CAREamics/careamics/blob/61c7b631cd8d38f41bca2fbb0675373227f893fc/src/careamics/lightning/modules/care_module.py#L193-L194
which is currently a bit of a hack a guess, but it allows the different normalisation methods.

File: src/careamics/lightning/modules/hdn_module.py (hunk 7 above)
Comment (melisande-c) [reply src/careamics/lightning/modules/hdn_module.py]: It is currently typed as `NDArray` in `ImageRegionData` but we should maybe change it to `NDArray | torch.Tensor`

File: src/careamics/lightning/modules/hdn_module.py (hunk 6 above)
Comment (melisande-c) [reply src/careamics/lightning/modules/hdn_module.py]: similar issue https://github.com/CAREamics/careamics/pull/994/changes#r3588854124

**Review by melisande-c — COMMENTED** [review]
Github not letting me add review without main comment

File: src/careamics/compat/config/configuration_factories.py
Comment (CatEek) [reply src/careamics/compat/config/configuration_factories.py:1]: done

File: src/careamics/config/architectures/lvae_config.py (hunk 3 above)
Comment (CatEek) [reply src/careamics/config/architectures/lvae_config.py]: Fixed

File: src/careamics/config/factories/hdn_factory.py
Comment (CatEek) [reply src/careamics/config/factories/hdn_factory.py:1]: Fixed

File: src/careamics/lightning/modules/hdn_module.py (hunk 5 above)
Comment (CatEek) [reply src/careamics/lightning/modules/hdn_module.py]: fixed in [17aa7cc](https://github.com/CAREamics/careamics/pull/994/commits/17aa7ccef0fc980b448429119e60828006673b6c)

File: src/careamics/lightning/modules/vae_lightning_module.py
Comment (CatEek) [reply src/careamics/lightning/modules/vae_lightning_module.py:1]: done

File: src/careamics/lightning/modules/hdn_module.py (hunk 6 above)
Comment (CatEek) [reply src/careamics/lightning/modules/hdn_module.py]: I believe HDN requires it, added explicit check [ee36159](https://github.com/CAREamics/careamics/pull/994/commits/ee36159e0166079194930d0614f27d835aae9430)

File: src/careamics/lightning/modules/microsplit_module.py (hunk 9 above)
Comment (CatEek) [reply src/careamics/lightning/modules/microsplit_module.py]: fixed in [6a7c8b5](https://github.com/CAREamics/careamics/pull/994/commits/6a7c8b55da1a686ec940f460d74099db51bdd52f)

File: src/careamics/lightning/modules/hdn_module.py (hunk 7 above)
Comment (CatEek) [reply src/careamics/lightning/modules/hdn_module.py]: added NDArray | torch.Tensor [6a7c8b5](https://github.com/CAREamics/careamics/pull/994/commits/6a7c8b55da1a686ec940f460d74099db51bdd52f)

File: src/careamics/lightning/modules/microsplit_module.py (hunk 11 above)
Comment (CatEek) [reply src/careamics/lightning/modules/microsplit_module.py]: changed, [12d4a84](https://github.com/CAREamics/careamics/pull/994/commits/12d4a84bcce6f18da6135c7881ffaadf85409759)

File: src/careamics/compat/config/configuration_factories.py (hunk 1 above)
Comment (jdeschamps) [reply src/careamics/compat/config/configuration_factories.py]: In `dev/microsplit_api` the "analytical_kl" is still mentioned in the docstring of `create_advanced_microsplit_config` and is available as a model parameter.  It should be kicked out of the comments and maybe even from the model parameters?

File: src/careamics/config/architectures/lvae_config.py (hunk 3 above)
Comment (jdeschamps) [reply src/careamics/config/architectures/lvae_config.py]: Fixed how? I see that the validation has been removed, the parameters are still there. Why was the validation irrelevant?

File: src/careamics/config/factories/hdn_factory.py
Comment (jdeschamps) [reply src/careamics/config/factories/hdn_factory.py:1]: Fixed how? Could you write some details because it is impossible to follow otherwise.

File: src/careamics/lightning/modules/hdn_module.py (hunk 6 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/hdn_module.py]: I see that the likelihood loss with noise model denormalize with explicit mean and std. The configuration should not allow a normalization that is incompatible. Can you open an issue?

File: src/careamics/lightning/modules/hdn_module.py (hunk 7 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/hdn_module.py]: > It is currently typed as `NDArray` in `ImageRegionData` but we should maybe change it to `NDArray | torch.Tensor`

We should open an issue

File: src/careamics/lightning/modules/microsplit_module.py [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+        )
+
+        return ImageRegionData(
+            data=denormalized_output,
+            source=x.source,
+            data_shape=output_data_shape,
+            dtype=x.dtype,
+            axes=x.axes,
+            region_spec=x.region_spec,
+            additional_metadata={},
```
Comment (jdeschamps) [inline src/careamics/lightning/modules/microsplit_module.py]: Still missing additional metadata forwarding.

File: src/careamics/lightning/modules/microsplit_module.py (hunk 10 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/microsplit_module.py:62]: Can you open an issue @CatEek ?

File: src/careamics/lightning/modules/hdn_module.py (hunk 7 above)
Comment (melisande-c) [reply src/careamics/lightning/modules/hdn_module.py]: Also technically typing the batch as `ImageRegionData` is slightly incorrect anyway since in reality the lightning module receives the collated version. I've just ignored typing issues that arise from this is the past, but if we wanted to be truly correct we would define a new class just for typing where the inner objects are Sequences of those present in `ImageRegionData`

File: src/careamics/config/architectures/lvae_config.py (hunk 3 above)
Comment (CatEek) [reply src/careamics/config/architectures/lvae_config.py]: Most likely that’s an artifact of how lvae Module was written. Technically we dont need 2 separate params Most likely, but that would fall into refactoring of lvae. Will create an issue

File: src/careamics/config/factories/hdn_factory.py
Comment (CatEek) [reply src/careamics/config/factories/hdn_factory.py:1]: Those are just merge conflicts

Comment (jdeschamps) [reply src/careamics/config/factories/hdn_factory.py:1]: I needed specifics... Now I had to spend the time investigating this branch and `dev/microsplit_api` to realize that this file seems to be gone from this PR. Probably because it was just a placeholder and your previous PR implemented the correct HDN factory.

File: src/careamics/lightning/modules/microsplit_module.py (hunk 12 above)
Comment (CatEek) [reply src/careamics/lightning/modules/microsplit_module.py]: fixed [5f37d13](https://github.com/CAREamics/careamics/pull/994/commits/5f37d13235d864b5ac7c1a541532916757040505)

File: src/careamics/compat/config/configuration_factories.py (hunk 1 above)
Comment (CatEek) [reply src/careamics/compat/config/configuration_factories.py]: removed [4905304](https://github.com/CAREamics/careamics/pull/994/commits/4905304b6250067f9d18f0f8f8085fbc5737f5f5)

**Review by jdeschamps — APPROVED** [review]
Looks good to me! @melisande-c ?

File: src/careamics/lightning/modules/microsplit_module.py [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+            `None` if the loss is NaN (so the caller can skip the batch).
+        """
+        return self.loss_func(
+            model_outputs=model_outputs,
+            targets=target,
+            config=self.config.loss,
+            noise_model=self.noise_model,
+            data_mean=self.data_mean,
+            data_std=self.data_std,
+        )
```
Comment (melisande-c) [inline src/careamics/lightning/modules/microsplit_module.py:187]: hmm probably the `self.data_mean` and `self.data_std` should also be taken from the dataset instead of the `set_data_stats` method. This means that for now MicroSplit also needs to be enforced to use `MeasStdNormalization`.

I am pretty sure these stats are only used to denomalize the data for the noise model, but I think we have discussed in the past whether the noise model can be trained on normalised data to avoid this.

**Review by melisande-c — COMMENTED** [review]
Sorry one last thing I didn't notice before

File: src/careamics/lightning/modules/microsplit_module.py (hunk 13 above)
Comment (CatEek) [reply src/careamics/lightning/modules/microsplit_module.py:187]: Done [91ecaea](https://github.com/CAREamics/careamics/pull/994/commits/91ecaead32c01a43380c48245480d943d2e62d72)

File: src/careamics/config/losses/loss_config.py (hunk 4 above)
Comment (veegalinova) [reply src/careamics/config/losses/loss_config.py:8]: It was used for kl loss annealing, as per discussions with the lab no one uses it as of now
